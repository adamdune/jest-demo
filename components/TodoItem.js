import PropTypes from "prop-types"

const TodoItem = ({ todo, index, onChangeCheckbox }) => (
  <div className="flex items-center cursor-pointer">
    <input
      type="checkbox"
      onChange={() => onChangeCheckbox(todo.isDone, index)}
      checked={todo.isDone}
      id={`checkbox/${index}`}
      className="cursor-pointer"
    />
    <div className="w-2" />
    <label
      htmlFor={`checkbox/${index}`}
      className={`${todo.isDone ? "line-through" : ""} cursor-pointer`}
    >
      {todo.title}
    </label>
  </div>
)

TodoItem.propTypes = {
  todo: PropTypes.shape({
    title: PropTypes.string.isRequired,
    isDone: PropTypes.bool.isRequired,
  }).isRequired,
  index: PropTypes.number.isRequired,
  onChangeCheckbox: PropTypes.func.isRequired,
}

export default TodoItem
