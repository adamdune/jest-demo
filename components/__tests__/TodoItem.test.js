import { fireEvent, render, screen } from "@testing-library/react"
import TodoItem from "../TodoItem.js"

describe("TodoItem", () => {
  it("while todo.isDone is true, label is rendered with a line-thorugh", () => {
    render(
      <TodoItem
        todo={{ title: "This is a todo", isDone: true }}
        onChangeCheckbox={() => {}}
        index={0}
      />,
    )

    const label = screen.getByText("This is a todo")

    expect(label).toHaveClass("line-through")
  })

  it("while todo.isDone is false, label is rendered without a line-thorugh", () => {
    render(
      <TodoItem
        todo={{ title: "This is a todo", isDone: false }}
        onChangeCheckbox={() => {}}
        index={0}
      />,
    )

    const label = screen.getByText("This is a todo")

    expect(label).not.toHaveClass("line-through")
  })

  it("when checkbox is clicked, onChangeCheckbox is called", () => {
    const callback = jest.fn()

    render(
      <TodoItem
        todo={{ title: "This is a todo", isDone: false }}
        onChangeCheckbox={callback}
        index={0}
      />,
    )

    const input = screen.getByRole("checkbox")

    expect(callback).toHaveBeenCalledTimes(0)
    fireEvent.click(input)
    expect(callback).toHaveBeenCalledTimes(1)
  })

  it("when label is clicked, onChangeCheckbox is called", () => {
    const callback = jest.fn()

    render(
      <TodoItem
        todo={{ title: "This is a todo", isDone: false }}
        onChangeCheckbox={callback}
        index={0}
      />,
    )

    const label = screen.getByText("This is a todo")

    expect(callback).toHaveBeenCalledTimes(0)
    fireEvent.click(label)
    expect(callback).toHaveBeenCalledTimes(1)
  })
})
