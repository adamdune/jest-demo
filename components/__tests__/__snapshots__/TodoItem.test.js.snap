// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoItem <TodoItem /> renders correctly when todo.isDone is false 1`] = `
<div
  className="flex items-center cursor-pointer"
  data-testid="TodoItem"
>
  <input
    checked={false}
    className="cursor-pointer"
    id="checkbox/0"
    onChange={[Function]}
    role="checkbox"
    type="checkbox"
  />
  <div
    className="w-2"
  />
  <label
    className=" cursor-pointer"
    htmlFor="checkbox/0"
  >
    Hello World
  </label>
</div>
`;

exports[`TodoItem <TodoItem /> renders correctly when todo.isDone is true 1`] = `
<div
  className="flex items-center cursor-pointer"
  data-testid="TodoItem"
>
  <input
    checked={true}
    className="cursor-pointer"
    id="checkbox/0"
    onChange={[Function]}
    role="checkbox"
    type="checkbox"
  />
  <div
    className="w-2"
  />
  <label
    className="line-through cursor-pointer"
    htmlFor="checkbox/0"
  >
    Lorem Ipsum
  </label>
</div>
`;
